test/unit/puppetlabs/puppetserver/certificate_authority_test.clj

```diff
@@ -510,6 +510,97 @@
+        crl-backup-path (str test-resources-dir "/update_crls/ca_crl.pem.bak")
+        new-crl-path (str test-resources-dir "/update_crls/new_root_crl.pem")
+        multiple-new-crls-path (str test-resources-dir "/update_crls/multiple_new_root_crls.pem")
+        multiple-newest-crls-path (str test-resources-dir "/update_crls/multiple_newest_root_crls.pem")
+        new-crls-for-multiple-certs-path (str test-resources-dir "/update_crls/new_crls_for_multiple_certs.pem")
+        new-and-unrelated-crls-path (str test-resources-dir "/update_crls/new_crls_and_unrelated_crls.pem")
+        unrelated-crls-path (str test-resources-dir "/update_crls/unrelated_crls.pem")
+        new-root-crl-chain-path (str test-resources-dir "/update_crls/chain_with_new_root.pem")
+        old-root-crl-path (str test-resources-dir "/update_crls/old_root_crl.pem")
+        delta-crl-path (str test-resources-dir "/update_crls/delta_crl.pem")
+        bad-inputs-and-error-msgs {multiple-newest-crls-path #"Could not determine newest CRL."
+                                   delta-crl-path #"Cannot support delta CRL."}]
+    (logutils/with-test-logging
+     (testing "a single newer CRL is used"
+       (let [incoming-crl-stream (testutils/pem-to-stream new-crl-path)]
+         (testutils/with-backed-up-crl crl-path crl-backup-path
+           (update-crls incoming-crl-stream crl-path cert-chain-path "foo")
+           (let [old-crls (utils/pem->crls crl-backup-path)
+                 new-crls (utils/pem->crls crl-path)
+                 old-number (utils/get-crl-number (last old-crls))
+                 new-number (utils/get-crl-number (last new-crls))]
+             (is (> new-number old-number))))))
+     (testing "the newest CRL given is used"
+       (let [incoming-crl-stream (testutils/pem-to-stream multiple-new-crls-path)]
+         (testutils/with-backed-up-crl crl-path crl-backup-path
+           (update-crls incoming-crl-stream crl-path cert-chain-path "foo")
+           (let [old-crls (utils/pem->crls crl-backup-path)
+                 new-crls (utils/pem->crls crl-path)
+                 old-number (utils/get-crl-number (last old-crls))
+                 new-number (utils/get-crl-number (last new-crls))]
+             (is (> new-number old-number))
+             (is (= 10 new-number))))))
+     (testing "multiple newer CRLs are used"
+       (let [incoming-crl-stream (testutils/pem-to-stream new-crls-for-multiple-certs-path)]
+         (testutils/with-backed-up-crl crl-path crl-backup-path
+           (update-crls incoming-crl-stream crl-path cert-chain-path "foo")
+           (let [old-crls (utils/pem->crls crl-backup-path)
+                 new-crls (utils/pem->crls crl-path)]
+             (doseq [new-crl new-crls
+                     old-crl old-crls]
+               (is (> (utils/get-crl-number new-crl)
+                      (utils/get-crl-number old-crl))))))))
+     (testing "unrelated CRLs are ignored while newer relevant CRLs are used"
```

Comment:
Is the point of this that incoming-crl-stream if filtered and/or sorted before being written?

Reply:
Yeah, I think the key is the filtering, plus using the key-id rather than the name. The fixture in question has four CRLs in it, two sets with the same names but different key-ids. It tests that we use the matching set and ignore the others.